tripal/tripal
# Update or add most remaining linking fields from issue #1381

# New Feature

### Closes #1381 
### Closes #1387
### Issue #1414 
### Issue #1550 

### Tripal Version: 4.alpha1

## Description
This PR will add all of the linking fields needed by core tripal, and updates a few existing ones to have linking capability.
For issue #1550 it will add the remaining missing fields from columns of the content type's base table to several content types, especially pub.
It will also add these linking fields to most content types through the yaml files in `tripal_chado/config/install/tripal.tripalfield_collection.xxx_chado.yml`

What is not done here is adding fields for the "complicated" content types. These are the cases where there are multiple content types derived from a single chado table. Four example, there are four content types from the `stock` table, they are all listed https://github.com/tripal/tripal_doc/issues/32#issuecomment-1837528225

Also, linking fields between two different entity groups are not yet added, for example stock (germplasm) <->feature (genetic or genomic)
Any table columns needing a timestamp are not here because the ChadoTimeDefault field does not yet exist (https://github.com/tripal/tripal/issues/1550#issuecomment-1623912019)

## Testing?
1. Build a docker on this branch
```
port=8080
git clone https://github.com/tripal/tripal.git testing-1381
cd testing-1381
git checkout tv4g1-issue1381-chado_linker_simple_fields
docker build --tag=tripaldocker:testing-1381 --build-arg drupalversion="10.1.x-dev" --build-arg chadoschema='chado' --file tripaldocker/Dockerfile-php8.2-pgsql13 ./
docker run --publish=${port}:80 --name=1381 -tid --volume=$(pwd):/var/www/drupal/web/modules/contrib/tripal tripalproject/tripaldocker:latest
docker exec testing-1381 service postgresql start
```
2. Go to "Page Structure" -> "+Import type collection"  -> Check all boxes and click "Import"
`drush trp-run-jobs --username=drupaladmin --root=/var/www/drupal/web`
There should of course be no errors.
3. Create some content for testing, this order lets you proceed nicely through the 15 tables use by core Tripal content:
Organism - Enter something for all the fields, and as you go down this list you will be able to link to the earlier content you just made. 
4. Contact
5. Publication - Make sure to put unique content in each of the publication fields to test them all. For ELocation and URL try something with a `https:` prefix to make sure it displays as a link
![20240110_ElocationURL](https://github.com/tripal/tripal/assets/8419404/eb9c85aa-009b-4b87-8ea6-179b2823dbf2)

7. Protocol
8. ArrayDesign - leave at least one of the later integer fields empty. Having a non-required integer field left blank tests the fix for issue #1387.
9. Study - specify both contact and publication when you create it. This tests the current version of chado storage - there should  be no error.
10. Assay
11. Analysis
12. Phylogenetic Tree
13. Project
14. Genetic Map
15. Germplasm accession (stock table)
16. Biological Sample
17. Gene (feature table)
18. DNA Library - Here the "Is Obsolete" field is a integer, and it is required. (Don't ask me why, chado defines it that way) Test using `0` to make sure zero is an integer! (Tests #1707)
![20240110_15tables](https://github.com/tripal/tripal/assets/8419404/42b20eb0-1b26-4327-adf4-0b2762ea1793)

19. Test as many other content types as you can stand, filling out all of the fields to make sure they work.
